src/adaptors.rs

```diff
@@ -714,3 +714,59 @@ impl<'a, I, F> Iterator for TakeWhileRef<'a, I, F> where
         (0, hi)
     }
 }
+
+/// An iterator to iterate through all the combinations of pairs in a **Clone**-able iterator.
+pub struct CombinatePair<I: Iterator> {
+    iter: I,
+    next_iter: Option<I>,
+    val: Option<I::Item>,
+}
+impl<I> CombinatePair<I> where I: Iterator {
+    /// Create a new **CombinatePair** from a clonable iterator.
+    pub fn new(iter: I) -> CombinatePair<I> {
+        CombinatePair { 
+            iter: iter, 
+            next_iter: None, 
+            val: None,
+        }
+    }
+}
+
+impl<I> Iterator for CombinatePair<I> where I: Iterator + Clone, I::Item: Clone{
+    type Item = (I::Item, I::Item);
+    fn next(&mut self) -> Option<Self::Item> {
+        // not having a value means we iterate once more throug the first iterator
+        if self.val.is_none() {
+            self.val = self.iter.next();
+            self.next_iter = Some(self.iter.clone());
+        }
+        // if its still none, we're out of values
+        if self.val.is_none() {
+            return None;
+        }
+        
+
+        let ret_ele = self.val.clone().unwrap();
+        match self.next_iter.as_mut().unwrap().next() {
+            Some(ref x) => {
+                return Some((ret_ele, x.clone()));
+            },
+            None => {
+                self.val = None;
+            }
+        }
+        // try again if we ran out of values in the second iterator
+        self.next()
+    }
```

Comment:
I'm sure we can write this in a way that .unwrap() isn't needed. For example `let elt = match self.val { Some(ref x) => x.clone(), None => return None };`.  Why not use the abbreviation `elt` like the other code too.